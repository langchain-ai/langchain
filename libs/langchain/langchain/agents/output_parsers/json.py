from __future__ import annotations

import logging
from typing import Union

from langchain_core.agents import AgentAction, AgentFinish
from langchain_core.exceptions import OutputParserException
from langchain_core.output_parsers.json import parse_json_markdown

from langchain.agents.agent import AgentOutputParser

logger = logging.getLogger(__name__)


class JSONAgentOutputParser(AgentOutputParser):
    """Parses tool invocations and final answers in JSON format.

    Expects output to be in one of two formats.

    If the output signals that an action should be taken,
    should be in the below format. This will result in an AgentAction
    being returned.

    ```
    {
      "action": "search",
      "action_input": "2+2"
    }
    ```

    If the output signals that a final answer should be given,
    should be in the below format. This will result in an AgentFinish
    being returned.

    ```
    {
      "action": "Final Answer",
      "action_input": "4"
    }
    ```
    """

    def parse(self, text: str) -> Union[AgentAction, AgentFinish]:
        try:
            response = parse_json_markdown(text)
            if isinstance(response, list):
                # gpt turbo frequently ignores the directive to emit a single action
                logger.warning("Got multiple action responses: %s", response)
                response = response[0]
            if response["action"] == "Final Answer":
                return AgentFinish({"output": response["action_input"]}, text)
            else:
                action_input = response.get("action_input", {})
                if action_input is None:
                    action_input = {}
                return AgentAction(response["action"], action_input, text)
        except Exception as e:
            raise OutputParserException(f"Could not parse LLM output: {text}") from e

    @property
    def _type(self) -> str:
        return "json-agent"
