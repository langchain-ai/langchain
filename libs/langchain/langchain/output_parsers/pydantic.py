"""Moved to langchain_core.output_parsers."""

from langchain_core.output_parsers import PydanticOutputParser

__all__ = ["PydanticOutputParser"]
