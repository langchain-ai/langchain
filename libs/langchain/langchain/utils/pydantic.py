from langchain_core.utils.pydantic import (  # type: ignore[deprecated]
    get_pydantic_major_version,
)

__all__ = ["get_pydantic_major_version"]
