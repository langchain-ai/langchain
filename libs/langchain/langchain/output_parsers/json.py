from langchain_core.output_parsers.json import SimpleJsonOutputParser
from langchain_core.utils.json import (parse_and_check_json_markdown,
                                       parse_json_markdown, parse_partial_json)

__all__ = [
    "SimpleJsonOutputParser",
    "parse_partial_json",
    "parse_json_markdown",
    "parse_and_check_json_markdown",
]
