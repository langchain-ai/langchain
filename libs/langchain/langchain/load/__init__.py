"""Serialization and deserialization."""

from langchain_core.load.dump import dumpd, dumps
from langchain_core.load.load import load, loads

__all__ = [
    "dumpd",
    "dumps",
    "load",
    "loads",
]
