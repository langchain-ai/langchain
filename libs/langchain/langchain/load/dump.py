from langchain_core.load.dump import default, dumpd, dumps

__all__ = ["default", "dumpd", "dumps"]
