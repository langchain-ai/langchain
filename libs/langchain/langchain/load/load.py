from langchain_core.load.load import Reviver, load, loads

__all__ = ["Reviver", "load", "loads"]
