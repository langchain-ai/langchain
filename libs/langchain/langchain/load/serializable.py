from langchain_core.load.serializable import (BaseSerialized, Serializable,
                                              SerializedConstructor,
                                              SerializedNotImplemented,
                                              SerializedSecret,
                                              to_json_not_implemented,
                                              try_neq_default)

__all__ = [
    "BaseSerialized",
    "SerializedConstructor",
    "SerializedSecret",
    "SerializedNotImplemented",
    "try_neq_default",
    "Serializable",
    "to_json_not_implemented",
]
