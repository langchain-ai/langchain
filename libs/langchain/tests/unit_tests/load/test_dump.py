"""Test for Serializable base class"""

import json
import os
from typing import Any, Dict, List
from unittest.mock import patch

import pytest
from langchain_core.load.dump import dumps
from langchain_core.load.serializable import Serializable
from pydantic import ConfigDict, Field, model_validator


class Person(Serializable):
    secret: str

    you_can_see_me: str = "hello"

    @classmethod
    def is_lc_serializable(cls) -> bool:
        return True

    @property
    def lc_secrets(self) -> Dict[str, str]:
        return {"secret": "SECRET"}

    @property
    def lc_attributes(self) -> Dict[str, str]:
        return {"you_can_see_me": self.you_can_see_me}


class SpecialPerson(Person):
    another_secret: str

    another_visible: str = "bye"

    @classmethod
    def get_lc_namespace(cls) -> List[str]:
        return ["my", "special", "namespace"]

    # Gets merged with parent class's secrets
    @property
    def lc_secrets(self) -> Dict[str, str]:
        return {"another_secret": "ANOTHER_SECRET"}

    # Gets merged with parent class's attributes
    @property
    def lc_attributes(self) -> Dict[str, str]:
        return {"another_visible": self.another_visible}


class NotSerializable:
    pass


def test_person(snapshot: Any) -> None:
    p = Person(secret="hello")
    assert dumps(p, pretty=True) == snapshot
    sp = SpecialPerson(another_secret="Wooo", secret="Hmm")
    assert dumps(sp, pretty=True) == snapshot
    assert Person.lc_id() == ["tests", "unit_tests", "load", "test_dump", "Person"]
    assert SpecialPerson.lc_id() == ["my", "special", "namespace", "SpecialPerson"]


def test_typeerror() -> None:
    assert (
        dumps({(1, 2): 3})
        == """{"lc": 1, "type": "not_implemented", "id": ["builtins", "dict"], "repr": "{(1, 2): 3}"}"""  # noqa: E501
    )


def test_person_with_kwargs(snapshot: Any) -> None:
    person = Person(secret="hello")
    assert dumps(person, separators=(",", ":")) == snapshot


def test_person_with_invalid_kwargs() -> None:
    person = Person(secret="hello")
    with pytest.raises(TypeError):
        dumps(person, invalid_kwarg="hello")


class TestClass(Serializable):
    my_favorite_secret: str = Field(alias="my_favorite_secret_alias")
    my_other_secret: str = Field()

    model_config = ConfigDict(
        populate_by_name=True,
    )

    @model_validator(mode="before")
    @classmethod
    def get_from_env(cls, values: Dict) -> Any:
        """Get the values from the environment."""
        if "my_favorite_secret" not in values:
            values["my_favorite_secret"] = os.getenv("MY_FAVORITE_SECRET")
        if "my_other_secret" not in values:
            values["my_other_secret"] = os.getenv("MY_OTHER_SECRET")
        return values

    @classmethod
    def is_lc_serializable(cls) -> bool:
        return True

    @classmethod
    def get_lc_namespace(cls) -> List[str]:
        return ["my", "special", "namespace"]

    @property
    def lc_secrets(self) -> Dict[str, str]:
        return {
            "my_favorite_secret": "MY_FAVORITE_SECRET",
            "my_other_secret": "MY_OTHER_SECRET",
        }


def test_aliases_hidden() -> None:
    test_class = TestClass(my_favorite_secret="hello", my_other_secret="world")  # type: ignore[call-arg]
    dumped = json.loads(dumps(test_class, pretty=True))
    expected_dump = {
        "lc": 1,
        "type": "constructor",
        "id": ["my", "special", "namespace", "TestClass"],
        "kwargs": {
            "my_favorite_secret": {
                "lc": 1,
                "type": "secret",
                "id": ["MY_FAVORITE_SECRET"],
            },
            "my_other_secret": {"lc": 1, "type": "secret", "id": ["MY_OTHER_SECRET"]},
        },
    }
    assert dumped == expected_dump
    # Check while patching the os environment
    with patch.dict(
        os.environ, {"MY_FAVORITE_SECRET": "hello", "MY_OTHER_SECRET": "world"}
    ):
        test_class = TestClass()  # type: ignore[call-arg]
        dumped = json.loads(dumps(test_class, pretty=True))

    # Check by alias
    test_class = TestClass(my_favorite_secret_alias="hello", my_other_secret="world")  # type: ignore[call-arg]
    dumped = json.loads(dumps(test_class, pretty=True))
    expected_dump = {
        "lc": 1,
        "type": "constructor",
        "id": ["my", "special", "namespace", "TestClass"],
        "kwargs": {
            "my_favorite_secret": {
                "lc": 1,
                "type": "secret",
                "id": ["MY_FAVORITE_SECRET"],
            },
            "my_other_secret": {"lc": 1, "type": "secret", "id": ["MY_OTHER_SECRET"]},
        },
    }
    assert dumped == expected_dump
