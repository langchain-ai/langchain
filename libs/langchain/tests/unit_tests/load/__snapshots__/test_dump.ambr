# serializer version: 1
# name: test_person
  '''
  {
    "lc": 1,
    "type": "constructor",
    "id": [
      "tests",
      "unit_tests",
      "load",
      "test_dump",
      "Person"
    ],
    "kwargs": {
      "secret": {
        "lc": 1,
        "type": "secret",
        "id": [
          "SECRET"
        ]
      },
      "you_can_see_me": "hello"
    }
  }
  '''
# ---
# name: test_person.1
  '''
  {
    "lc": 1,
    "type": "constructor",
    "id": [
      "my",
      "special",
      "namespace",
      "SpecialPerson"
    ],
    "kwargs": {
      "secret": {
        "lc": 1,
        "type": "secret",
        "id": [
          "SECRET"
        ]
      },
      "you_can_see_me": "hello",
      "another_secret": {
        "lc": 1,
        "type": "secret",
        "id": [
          "ANOTHER_SECRET"
        ]
      },
      "another_visible": "bye"
    }
  }
  '''
# ---
# name: test_person_with_kwargs
  '{"lc":1,"type":"constructor","id":["tests","unit_tests","load","test_dump","Person"],"kwargs":{"secret":{"lc":1,"type":"secret","id":["SECRET"]},"you_can_see_me":"hello"}}'
# ---
